// featureSync/index.js
// Part of <https://miracle.systems/p/walkner-xiconf> licensed under <CC BY-NC-SA 4.0>

'use strict';

var path = require('path');
var os = require('os');
var exec = require('child_process').exec;
var step = require('h5.step');

exports.DEFAULT_CONFIG = {
  settingsId: 'settings',
  rmdirCmd: 'rmdir /S /Q "{dir}"',
  copyCmd: 'xcopy /I /Q /Y "{src}" "{dst}"'
};

exports.start = function startFeatureSyncModule(app, module)
{
  var timer = null;
  var syncing = false;

  module.isSyncing = function() { return syncing; };

  module.sync = sync;

  app.broker.subscribe('app.started', scheduleSync);

  function scheduleSync()
  {
    var settings = app[module.config.settingsId];

    if (!settings)
    {
      return module.warn("Not syncing: no settings module.");
    }

    var delay = parseInt(settings.get('backupDelay'), 10) * 1000;

    if (isNaN(delay) || delay < 1000)
    {
      return module.warn("Not syncing: invalid backup delay.");
    }

    timer = setTimeout(sync, delay);
  }

  function sync(done)
  {
    var settings = app[module.config.settingsId];

    if (!settings)
    {
      return done && done(new Error('NO_SETTINGS'));
    }

    if (syncing)
    {
      return done && done(new Error('IN_PROGRESS'));
    }

    var backupPath = settings.get('backupPath') === 1 ? 1 : 2;
    var srcPath = backupPath === 1 ? settings.get('featurePath1') : settings.get('featurePath2');
    var dstPath = backupPath === 1 ? settings.get('featurePath2') : settings.get('featurePath1');
    var tmpPath = path.join(os.tmpdir(), 'walkner-xiconf-features');

    if (typeof srcPath !== 'string' || !srcPath.length
      || typeof dstPath !== 'string' || !dstPath.length)
    {
      if (done)
      {
        return done(new Error('INVALID_SETTINGS'));
      }

      return module.warn("Not syncing: invalid paths.");
    }

    if (srcPath.length)
    {
      srcPath = srcPath.replace(/(\/|\\)$/, '') + '/';
    }

    if (dstPath.length)
    {
      dstPath = dstPath.replace(/(\/|\\)$/, '');
    }

    if (timer !== null)
    {
      clearTimeout(timer);
      timer = null;
    }

    syncing = true;

    module.info("Started syncing...");

    if (done)
    {
      done(null, {
        tmpPath: tmpPath,
        srcPath: srcPath,
        dstPath: dstPath
      });
    }

    app.broker.publish('featureSync.started');

    step(
      function removeTmpDirStep()
      {
        module.debug("Removing the tmp dir...");

        this.cmd = module.config.rmdirCmd.replace('{dir}', tmpPath);

        exec(this.cmd, this.next());
      },
      function copySrcToTmpStep(err)
      {
        if (err)
        {
          module.debug(err, "Failed to remove the tmp dir.", {cmd: this.cmd});
        }

        module.debug("Copying the src dir to the tmp dir...");

        this.cmd = module.config.copyCmd
          .replace('{src}', srcPath)
          .replace('{dst}', tmpPath);

        exec(this.cmd, this.next());
      },
      function removeDstDirStep(err)
      {
        if (err)
        {
          module.error(err, "Failed to copy the src dir to the tmp dir.", {cmd: this.cmd});

          return this.skip(new Error('SRC_TO_TMP'));
        }

        module.debug("Removing the dst dir...");

        this.cmd = module.config.rmdirCmd.replace('{dir}', dstPath);

        exec(this.cmd, this.next());
      },
      function copyTmpToDstStep(err)
      {
        if (err)
        {
          module.error(err, "Failed to remove the dst dir.", {cmd: this.cmd});
        }

        module.debug("Copying the tmp dir to the dst dir...");

        this.cmd = module.config.copyCmd
          .replace('{src}', tmpPath)
          .replace('{dst}', dstPath);

        exec(this.cmd, this.next());
      },
      function removeTmpDirStep(err)
      {
        if (err)
        {
          module.error(err, "Failed to copy the tmp dir to the dst dir.", {cmd: this.cmd});

          return this.skip(new Error('TMP_TO_DST'));
        }

        module.debug("Removing the tmp dir...");

        var cmd = module.config.rmdirCmd.replace('{dir}', tmpPath);
        var next = this.next();

        exec(cmd, function(err)
        {
          if (err)
          {
            module.error(err, "Failed to remove the tmp dir.", {cmd});
          }

          next();
        });
      },
      function finalizeStep(err)
      {
        if (err)
        {
          module.warn(err, "Failed to sync the feature files.");

          app.broker.publish('featureSync.finished', {error: {message: err.message}});
        }
        else
        {
          module.info("Synced the feature files!");

          app.broker.publish('featureSync.finished', {});
        }

        syncing = false;
      }
    );
  }
};
