// updater/index.js
// Part of <https://miracle.systems/p/walkner-xiconf> licensed under <CC BY-NC-SA 4.0>

'use strict';

var os = require('os');
var path = require('path');
var url = require('url');
var fs = require('fs');
var exec = require('child_process').exec;
var util = require('util');
var _ = require('lodash');
var request = require('request');
var step = require('h5.step');
var semver = require('semver');

exports.DEFAULT_CONFIG = {
  settingsId: 'settings',
  unzipExe: 'unzip.exe'
};

exports.start = function startUpdaterModule(app, updaterModule)
{
  var settings = app[updaterModule.config.settingsId];

  var isCheckingUpdates = false;
  var isProgrammingInProgress = false;

  app.broker.subscribe('programmer.stateChanged', onProgrammerStateChanged);
  app.broker.subscribe('updater.checkRequested', update);

  function onProgrammerStateChanged(message)
  {
    if (message.inProgress !== undefined)
    {
      isProgrammingInProgress = message.inProgress;
    }
  }

  function update()
  {
    if (isCheckingUpdates)
    {
      return;
    }

    var updatesUrl = url.parse(settings.get('remoteServer'));

    if (_.isEmpty(updatesUrl.hostname))
    {
      return updaterModule.warn("Not updating: no remote server.");
    }

    if (_.isEmpty(updatesUrl.protocol))
    {
      updatesUrl.prototcol = 'http:';
    }

    var path = '/xiconf/updates/' + app.options.version;

    _.assign(updatesUrl, {
      path: path,
      pathname: path,
      query: '',
      search: '',
      hash: ''
    });

    updatesUrl = url.format(updatesUrl);

    checkUpdate(updatesUrl);
  }

  function checkUpdate(updatesUrl)
  {
    isCheckingUpdates = true;

    updaterModule.debug("Checking updates...", {updatesUrl});

    request
      .get(updatesUrl)
      .on('error', function(err)
      {
        isCheckingUpdates = false;

        updaterModule.error(err, "Failed to check updates.");
      })
      .on('response', function(res)
      {
        if (res.statusCode === 204)
        {
          isCheckingUpdates = false;

          return updaterModule.info("No updates!");
        }

        if (res.statusCode === 200)
        {
          updaterModule.debug("Downloading the update archive...");

          var updateZipPath = path.join(os.tmpdir(), 'XICONF_UPDATE-' + Math.round(Math.random() * 99999999) + '.zip');

          res.on('end', unzipUpdate.bind(null, updateZipPath));
          res.pipe(fs.createWriteStream(updateZipPath));

          return;
        }

        isCheckingUpdates = false;

        updaterModule.warn("Received an invalid status code.", {responseStatusCode: res.statusCode});
      });
  }

  function unzipUpdate(updateZipPath)
  {
    var updateDirPath = path.join(app.options.rootPath, '..', 'update');

    step(
      function rmdirStep()
      {
        updaterModule.debug("Removing old update directory...");

        var cmd = util.format(
          process.platform === 'win32' ? 'RD /S /Q "%s"' : 'rm -rf "%s"',
          updateDirPath
        );

        exec(cmd, {timeout: 15000}, this.next());
      },
      function unzipStep()
      {
        updaterModule.debug("Unzipping the update archive...");

        const cmd = updaterModule.config.unzipExe.includes('7z')
          ? `"${updaterModule.config.unzipExe}" x -y -o"${updateDirPath}" "${updateZipPath}"`
          : `"${updaterModule.config.unzipExe}" "${updateZipPath}" -d "${updateDirPath}"`;

        exec(cmd, {timeout: 60000}, this.next());
      },
      function checkStep(err)
      {
        fs.unlink(updateZipPath, function() {});

        if (err)
        {
          isCheckingUpdates = false;

          return updaterModule.error(err, "Failed to unzip the update.");
        }

        var oldPackageJson;
        var newPackageJson;

        try
        {
          oldPackageJson = require(path.join(app.options.rootPath, '..', 'package.json'));
          newPackageJson = findNewPackageJson(updateDirPath);

          if (newPackageJson.name !== oldPackageJson.name
            || !semver.gt(newPackageJson.version, oldPackageJson.version))
          {
            throw new Error(util.format(
              "invalid package.json: old %s v%s, new %s v%s",
              oldPackageJson.name,
              oldPackageJson.version,
              newPackageJson.name,
              newPackageJson.version
            ));
          }
        }
        catch (err)
        {
          isCheckingUpdates = false;

          return updaterModule.error(err, "Failed to require the package.json.");
        }

        setImmediate(installUpdate, updateDirPath, newPackageJson);
      }
    );
  }

  function installUpdate(updateDirPath, packageJson)
  {
    if (isProgrammingInProgress)
    {
      updaterModule.debug("Delaying update...", {newVersion: packageJson.version});

      return app.broker
        .subscribe('programmer.finished', installUpdate.bind(null, updateDirPath, packageJson))
        .setLimit(1);
    }

    updaterModule.debug("Updating...", {oldVersion: app.options.version, newVersion: packageJson.version});

    app.broker.publish('updater.updating', {
      fromVersion: app.options.version,
      toVersion: packageJson.version
    });

    setTimeout(function()
    {
      try
      {
        require(path.join(updateDirPath, 'install.js'))(app, updaterModule, packageJson);
      }
      catch (err)
      {
        updaterModule.error(err, "Failed to install update");

        process.exit(1); // eslint-disable-line no-process-exit
      }
    }, 1337 * 2);
  }

  function findNewPackageJson(updateDirPath)
  {
    try
    {
      return require(path.join(updateDirPath, 'data', 'app', 'package.json'));
    }
    catch (err)
    {
      try
      {
        return require(path.join(updateDirPath, 'data', 'package.json'));
      }
      catch (err)
      {
        return require(path.join(updateDirPath, 'package.json'));
      }
    }
  }
};
