// h5-express/errorHandlerMiddleware.js
// Part of <https://miracle.systems/p/walkner-snf> licensed under <CC BY-NC-SA 4.0>

'use strict';

const _ = require('lodash');

module.exports = (expressModule, options) =>
{
  if (!options)
  {
    options = {};
  }

  // Based on https://github.com/expressjs/errorhandler
  return (err, req, res, next) => // eslint-disable-line no-unused-vars
  {
    if (_.includes(expressModule.config.ignoredErrorCodes, err.code))
    {
      return;
    }

    if (err.status || err.statusCode)
    {
      res.statusCode = err.status || err.statusCode;
    }

    if (res.statusCode < 400)
    {
      res.statusCode = 500;
    }

    if (typeof err === 'string')
    {
      err = {message: err, stack: null};
    }
    else if (err && err.name === 'ValidationError')
    {
      _.forEach(err.errors, (validationError) =>
      {
        err.message += '\n  - ' + validationError;
      });
    }

    const login = req.session && req.session.user
      ? req.session.user.login
      : 'guest';

    expressModule.warn(err, {
      req: {
        method: req.method,
        url: req.url,
        ip: req.ip,
        headers: req.headers
      },
      res: {
        statusCode: res.statusCode
      },
      user: {
        login
      }
    });

    if (!res.connection || !res.connection.writable || res.headersSent)
    {
      return;
    }

    const accept = req.headers.accept || '';

    if (accept.indexOf('html') !== -1)
    {
      res.render('error', {
        title: options.title || 'express',
        statusCode: res.statusCode,
        stack: prepareStack(options.basePath, err).reverse(),
        error: err.message.replace(/^Error: /, '')
      });

      return;
    }

    if (accept.indexOf('json') !== -1)
    {
      const error = {
        message: err.message,
        stack: err.stack
      };

      Object.keys(err).forEach(k => error[k] = err[k]);

      res.type('json');
      res.end(JSON.stringify({error}, getCircularReplacer()));

      return;
    }

    res.setHeader('Content-Type', 'text/plain');
    res.end(err.stack || err.message);
  };
};

function getCircularReplacer()
{
  const seen = new WeakSet();

  return (key, value) =>
  {
    if (typeof value === 'object' && value !== null)
    {
      if (seen.has(value))
      {
        return;
      }

      seen.add(value);
    }

    return value;
  };
}

function prepareStack(basePath, err)
{
  const stack = (err.stack || '').split('\n').slice(1);

  if (stack.length === 0)
  {
    return [];
  }

  let no = stack.length;

  return stack.map(function(stack)
  {
    let matches = stack.match(/at (.*?) \((.*?):([0-9]+):([0-9]+)\)/);

    if (matches !== null)
    {
      return {
        no: no--,
        func: matches[1],
        path: matches[2],
        file: extractFile(basePath, matches[2]),
        line: matches[3],
        column: matches[4]
      };
    }

    matches = stack.match(/at (.*?):([0-9]+):([0-9]+)/);

    if (matches !== null)
    {
      return {
        no: no--,
        func: '?',
        path: matches[1],
        file: extractFile(basePath, matches[1]),
        line: matches[2],
        column: matches[3]
      };
    }

    return {
      no: no--,
      unknown: stack
    };
  });
}

function extractFile(basePath, path)
{
  if (!basePath || path.toLowerCase().indexOf(basePath.toLowerCase()) !== 0)
  {
    return path;
  }

  return '.' + path.substr(basePath.length);
}
