// h5-express/MongoStore.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const util = require('util');
const Store = require('express-session').Store;

module.exports = MongoStore;

/**
 * @constructor
 * @param {Db} db
 * @param {MongoStore.Options} [options]
 */
function MongoStore(db, options)
{
  options = options || {};

  Store.call(this, options);

  /**
   * @private
   * @type {string}
   */
  this.collectionName = options.collectionName || MongoStore.Options.collectionName;

  /**
   * @private
   * @type {number}
   */
  this.gcInterval = (options.gcInterval || MongoStore.Options.gcInterval) * 1000;

  /**
   * @private
   * @type {number}
   */
  this.defaultExpirationTime
    = (options.defaultExpirationTime || MongoStore.Options.defaultExpirationTime) * 1000;

  /**
   * @private
   * @type {number}
   */
  this.touchChance = options.touchChance || MongoStore.Options.touchChance;

  /**
   * @private
   * @type {number}
   */
  this.touchInterval = options.touchInterval || MongoStore.Options.touchInterval;

  /**
   * @private
   * @type {boolean}
   */
  this.cacheInMemory = options.cacheInMemory === true;

  /**
   * @private
   * @type {Map<string, object>}
   */
  this.cache = new Map();

  /**
   * @private
   * @type {number|null}
   */
  this.gcTimer = null;

  /**
   * @private
   * @type {Db}
   */
  this.db = db;

  /**
   * @private
   * @type {function}
   */
  this.onOpen = this.onOpen.bind(this);

  /**
   * @private
   * @type {function}
   */
  this.onClose = this.onClose.bind(this);

  this.db.on('reconnect', this.onOpen);
  this.db.on('close', this.onClose);

  this.scheduleGc();
}

/**
 * @type {object}
 */
MongoStore.Options = {
  /**
   * @type {string}
   */
  collectionName: 'sessions',

  /**
   * @type {number}
   */
  gcInterval: 600,

  /**
   * @type {number}
   */
  defaultExpirationTime: 3600 * 24 * 14,

  /**
   * @type {number}
   */
  touchChance: 0.25,

  /**
   * @type {number}
   */
  touchInterval: 0,

  /**
   * @type {boolean}
   */
  cacheInMemory: false
};

util.inherits(MongoStore, Store);

MongoStore.prototype.touch = function(sid, session, done)
{
  const now = Date.now();

  if ((this.touchInterval > 0 && (now - session.updatedAt) < this.touchInterval)
    || (this.touchChance > 0 && Math.random() > this.touchChance))
  {
    return done(null);
  }

  const sessions = this.collection();
  const expires = Date.parse(session.cookie.expires) || (Date.now() + this.defaultExpirationTime);
  const update = {expires, updatedAt: now};

  if (this.cache.has(sid))
  {
    Object.assign(this.cache.get(sid), update);
  }

  sessions.updateOne({_id: sid}, {$set: update}, err =>
  {
    if (done)
    {
      done(err);
    }
  });
};

/**
 * @param {string} sid
 * @param {function} done
 */
MongoStore.prototype.get = function(sid, done)
{
  const store = this;

  if (store.cache.has(sid))
  {
    handleGet(null, store.cache.get(sid));
  }
  else
  {
    store.collection().findOne({_id: sid}, handleGet);
  }

  function handleGet(err, doc)
  {
    if (err)
    {
      return done(err);
    }

    if (!doc)
    {
      store.cache.delete(sid);

      return done();
    }

    const session = typeof doc.data === 'string' ? JSON.parse(doc.data) : doc.data;
    const expires = typeof session.cookie.expires === 'string'
      ? new Date(session.cookie.expires)
      : session.cookie.expires;

    if (!expires || Date.now() < expires)
    {
      session.updatedAt = doc.updatedAt;

      if (store.cacheInMemory && !store.cache.has(sid))
      {
        store.cache.set(sid, doc);
      }

      return done(null, session);
    }

    return store.destroy(sid, done);
  }
};

/**
 * @param {string} sid
 * @param {Session} session
 * @param {function} [done]
 */
MongoStore.prototype.set = function(sid, session, done)
{
  const sessions = this.collection();
  const now = Date.now();
  const doc = {
    _id: sid,
    updatedAt: now,
    expires: Date.parse(session.cookie.expires),
    data: session
  };

  if (isNaN(doc.expires))
  {
    doc.expires = now + this.defaultExpirationTime;
  }

  if (this.cacheInMemory)
  {
    this.cache.set(sid, doc);
  }

  sessions.replaceOne({_id: sid}, doc, {upsert: true}, err =>
  {
    if (done)
    {
      done(err);
    }
  });
};

/**
 * @param {string} sid
 * @param {function} [done]
 */
MongoStore.prototype.destroy = function(sid, done)
{
  this.cache.delete(sid);

  this.collection().deleteOne({_id: sid}, err =>
  {
    if (done)
    {
      done(err);
    }
  });
};

/**
 * @param {function} [done]
 */
MongoStore.prototype.clear = function(done)
{
  this.cache.clear();

  this.collection().drop(err =>
  {
    if (done)
    {
      done(err);
    }
  });
};

/**
 * @param {function} done
 */
MongoStore.prototype.length = function(done)
{
  this.collection().countDocuments({}, done);
};

/**
 * @param {function} [done]
 */
MongoStore.prototype.gc = function(done)
{
  this.cache.clear();

  this.collection().deleteMany({expires: {$lte: Date.now()}}, err =>
  {
    if (done)
    {
      done(err);
    }
  });
};

MongoStore.prototype.destruct = function()
{
  this.cache.clear();

  this.clearGcTimer();

  this.db.removeListener('open', this.onOpen);
  this.db.removeListener('close', this.onClose);
  this.db = null;
};

/**
 * @private
 * @returns {Object}
 */
MongoStore.prototype.collection = function()
{
  return this.db.collection(this.collectionName);
};

/**
 * @private
 */
MongoStore.prototype.onOpen = function()
{
  this.scheduleGc();
};

/**
 * @private
 */
MongoStore.prototype.onClose = function()
{
  this.clearGcTimer();
};

/**
 * @private
 */
MongoStore.prototype.clearGcTimer = function()
{
  if (this.gcTimer !== null)
  {
    clearTimeout(this.gcTimer);
    this.gcTimer = null;
  }
};

/**
 * @private
 */
MongoStore.prototype.scheduleGc = function()
{
  this.gcTimer = setTimeout(
    () =>
    {
      this.gcTimer = null;
      this.gc(this.scheduleGc.bind(this));
    },
    this.gcInterval
  );
};
