// programmer/glp2/index.js
// Part of <https://miracle.systems/p/walkner-xiconf> licensed under <CC BY-NC-SA 4.0>

'use strict';

var constants = require('./constants');
var util = require('./util');

exports.CHR = util.CHR;
exports.prettifyBuffer = util.prettifyBuffer;

exports.ResponseType = constants.ResponseType;
exports.FaultStatus = constants.FaultStatus;
exports.OnOff = constants.OnOff;
exports.TransmittingMode = constants.TransmittingMode;
exports.GoSwitchAtVisualCheck = constants.GoSwitchAtVisualCheck;
exports.PeDisplay = constants.PeDisplay;
exports.PeSetValues = constants.PeSetValues;
exports.StartSafetySwitch = constants.StartSafetySwitch;
exports.CnfFunction = constants.CnfFunction;
exports.SelfTest = constants.SelfTest;
exports.HvProcess = constants.HvProcess;
exports.AddProtocolNumber = constants.AddProtocolNumber;
exports.RemoteControl = constants.RemoteControl;
exports.OrderInputs = constants.OrderInputs;
exports.TestStepDisplay = constants.TestStepDisplay;
exports.ActualValueData = constants.ActualValueData;

exports.BusyTesterError = require('./BusyTesterError');
exports.InvalidChecksumError = require('./InvalidChecksumError');
exports.InvalidParametersError = require('./InvalidParametersError');
exports.InvalidResponseError = require('./InvalidResponseError');
exports.NoConnectionError = require('./NoConnectionError');
exports.ResponseTimeoutError = require('./ResponseTimeoutError');

exports.Master = require('./Master');

exports.ActualValuesResponse = require('./ActualValuesResponse');
exports.DeviceOptionsResponse = require('./DeviceOptionsResponse');
exports.InterimActualValuesResponse = require('./InterimActualValuesResponse');
exports.OrderDataDemandResponse = require('./OrderDataDemandResponse');

exports.ProgramStep = require('./ProgramStep');
exports.PeTest = require('./PeTest');
exports.IsoTest = require('./IsoTest');
exports.FctTest = require('./FctTest');
exports.VisTest = require('./VisTest');

exports.TestResult = require('./TestResult');
exports.DefaultTestResult = require('./DefaultTestResult');
exports.HvTestResult = require('./HvTestResult');
exports.VisTestResult = require('./VisTestResult');
